packages/core/src/contrib/Constraints.ts

```diff
@@ -229,6 +229,53 @@ export const objDict = {
   nearScalar: (c: any, goal: any) => {
     return squared(sub(constOfIf(c), constOfIf(goal)));
   },
+  /**
+   * try to make distance between a point and a segment `s1` = padding.
+   */
+  closestPtToSeg: ([, s1]: [string, any], point: VarAD[], padding: VarAD) => {
+    return equalHard(
+      ops.vdist(
+        closestPt_PtSeg(point, [s1.start.contents, s1.end.contents]),
+        point
+      ),
+      padding
+    );
+  },
+  /**
+   * Repel the angle between the p1-p0 and p1-p2 away from 0 and 180 degrees. NOTE: angles between 10 and 170 degrees are considered satisfied.
+   */
+  nonDegenerateAngle: (
+    [, p0]: [string, any],
+    [, p1]: [string, any],
+    [, p2]: [string, any],
+    strength = 20
+  ) => {
+    if (every([p0, p1, p2].map((props) => props["center"]))) {
+      const c0 = fns.center(p0);
+      const c1 = fns.center(p1);
+      const c2 = fns.center(p2);
+
+      const l1 = ops.vsub(c0, c1);
+      const l2 = ops.vsub(c2, c1);
+
+      // angles between 10 and 170 degrees do not need to be pushed
+      return ifCond(
+        lt(
+          absVal(ops.vdot(ops.vnormalize(l1), ops.vnormalize(l2))),
```

Comment:
factor the cosine calculation into a new local variable and re-use it for L264 and L270